pkg/kn/plugin/context_sharing.go

```diff
@@ -0,0 +1,181 @@
+}
+
+type ContextDataConsumer interface {
+	// ExecuteWithContextData executes the plugin with the given args much like
+	// Execute() but with an additional argument that holds the ContextData
+	ExecuteWithContextData(args []string, data ContextData) error
+}
+
+//--TYPES--
+
+var ctxManager *ContextDataManager
+
+type ContextDataManager struct {
+	ContextData map[string]ContextData `json:"contextData"`
+	Manifests   map[string]Manifest    `json:"manifests"`
+}
+
+func NewContextManager() (*ContextDataManager, error) {
+	if ctxManager == nil {
+		var err error
+		//file, err := os.Open("~/.config/kn/context.json")
+		//if err != nil {
+		//	return nil, err
+		//}
+		//TODO: mock data to test things out
+		mockData := `
+{
+	"contextData": {
+		"default": {
+			"service": "hello",
+			"namespace": "default"
+		}
+	},
+	"manifests": {
+		"kn-service-log": {
+			"path": "/usr/local/bin/kn-service/log",
+			"hasManifest": true,
+			"description": "Show logs of a service",
+			"consumesKeys": ["service", "namespace"]
+		}
+	}
+}
+`
+		mockReader := strings.NewReader(mockData)
+		decoder := json.NewDecoder(mockReader)
+		contextData := &ContextDataManager{}
+		err = decoder.Decode(&contextData)
+		if err != nil {
+			return nil, err
+		}
+		ctxManager = contextData
+	}
+	return ctxManager, nil
+}
+
+// GetContext returns context data by key
+func (c *ContextDataManager) GetContext(key string) ContextData {
+	return c.ContextData[key]
+}
+
+// GetDefault returns default context data
+func (c *ContextDataManager) GetDefault() ContextData {
+	return c.GetContext("default")
+}
+
+// GetConsumesKeys returns array of keys consumed by plugin
+func (c *ContextDataManager) GetConsumesKeys(pluginName string) []string {
+	return c.Manifests[pluginName].ConsumesContextDataKeys
+}
+
+// GetProducesKeys returns array of keys produced by plugin
+func (c *ContextDataManager) GetProducesKeys(pluginName string) []string {
+	return c.Manifests[pluginName].ProducesContextDataKeys
+}
+
+// FetchManifests it tries to retrieve manifest from both inlined and external plugins
+func (c *ContextDataManager) FetchManifests(pluginManager *Manager) error {
+	plugins, err := pluginManager.ListPlugins()
+	if err != nil {
+		return err
+	}
+	for _, p := range plugins {
+		// Add new plugins only
+		if _, exists := c.Manifests[p.Name()]; !exists {
+			manifest := &Manifest{}
+			if p.Path() != "" {
+				// Fetch from external plugin
+				if m := fetchExternalManifest(p); m != nil {
+					manifest = m
+				}
+			} else {
+				// Fetch from internal plugin
+				if pwm, ok := p.(PluginWithManifest); ok {
+					manifest = pwm.GetManifest()
+				}
+			}
+			// Add manifest to map
+			c.Manifests[p.Name()] = *manifest
+		}
+	}
+	return nil
+}
+
+// TODO: We should cautiously execute external binaries
+// fetchExternalManifest returns Manifest from external plugin by exec `$plugin manifest get`
+func fetchExternalManifest(p Plugin) *Manifest {
```

Comment:
That's a bit tricky part. Executing binary half blind if it's desired one even?

Reply:
Preview of manifest flag: https://github.com/knative-extensions/kn-plugin-service-log/commit/c214f72c49c7da7179a6869a9af8e9bc38785f3f

That can be a standalone cmd. I.e. `kn service log manifest`, but it kind of makes single purpose plugins grow the list of cmds. :)

Reply:
Indeed, the tricky part is not blindly calling each plugin in the path somehow, if possible. Of course *if* a plugin is in the path, it can be called anyway (maybe by accident), but chances are higher that a malicious plugin is called when it is called by itself. A simple solution could be to keep track of which plugins have been called explicitly and only can them when picking up all manifests (which might be hard to maintain, though ...)